File: packages/guidance/src/policies/noMutableVariableDeclarations.ts
import { Function } from "effect"
import type { Match } from "@better-typescript/matchers/matcher/data"
import { oneFinding } from "@better-typescript/core/engine/policy"
import { defineBuiltinPolicy } from "../definePolicy.js"
import {
  noMutableVariableDeclarationsMatcher,
  type NoMutableVariableDeclarationsFact
} from "@better-typescript/matchers/builtins/noMutableVariableDeclarations"

const hint =
  "Declare multiple const values to represent each state instead of mutating a single " +
  "variable, and use immutable values that are not reassigned. When the value must " +
  "genuinely evolve over time (a module-level counter, a cell shared across " +
  "closures), hold it in a Ref inside the Effect runtime instead of a let binding."

const noMutableVariableDeclarationsFindings = (match: Match<NoMutableVariableDeclarationsFact>) =>
  oneFinding(
    match.target,
    `Avoid declaring mutable variables with ${match.fact.kind}.`,
    hint,
    undefined
  )

export const noMutableVariableDeclarations = defineBuiltinPolicy(
  "no-mutable-variable-declarations",
  noMutableVariableDeclarationsMatcher,
  Function.constant(noMutableVariableDeclarationsFindings)
)

File: internal/rules/no_mutable_variable_declarations/no_mutable_variable_declarations.go
package no_mutable_variable_declarations

import (
	"fmt"
	"github.com/andrueandersoncs/better-typescript/internal/rule"
	"github.com/andrueandersoncs/typescript-go/ast"
)

const help = "Application code should declare multiple const values to represent each state instead of mutating one variable. When shared state genuinely evolves over time, use a Ref inside the Effect runtime instead of a let binding. An owned library kernel may use local mutation only under explicit project policy; this rule does not infer that exception."

var Rule = rule.Rule{Name: "no-mutable-variable-declarations", Run: run}

func run(ctx rule.RuleContext, _ any) rule.RuleListeners {
	return rule.RuleListeners{ast.KindVariableDeclarationList: func(node *ast.Node) {
		kind := ""
		switch node.Flags & ast.NodeFlagsBlockScoped {
		case 0:
			kind = "var"
		case ast.NodeFlagsLet:
			kind = "let"
		}
		if kind != "" {
			ctx.ReportNode(node, rule.RuleMessage{Id: "no-mutable-variable-declarations", Description: fmt.Sprintf("Avoid declaring mutable variables with %s.", kind), Help: help})
		}
	}}
}
